File: lena-config/src/test/java/org/riversoforion/lena/config/SimpleConfigurationSourceTest.java
/*
 * Copyright (c) 2024-2025. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class SimpleConfigurationSourceTest {

    @Test
    @DisplayName("getValue delegates to resolvers")
    void getValue_DelegatesToResolvers(@Mock NameResolver names, @Mock ValueResolver values) {

        Namespace root = Namespace.root();
        Name thisProp = Name.of("this.prop");
        when(names.resolveName(root, thisProp)).thenReturn("THIS_PROP");
        when(values.resolveValue("THIS_PROP")).thenReturn(Optional.of("some_value"));

        SimpleConfigurationSource source = new SimpleConfigurationSource(names, values);
        Optional<String> value = source.getValue(root, thisProp);

        assertThat(value).contains("some_value");
        verify(names).resolveName(root, thisProp);
        verify(values).resolveValue("THIS_PROP");
    }
}

File: lena-config/src/commonTest/kotlin/org/riversoforion/lena/config/SimpleConfigurationSourceTest.kt
/*
 * Copyright (c) 2024-2026. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config

import kotlin.test.Test
import kotlin.test.assertEquals

class SimpleConfigurationSourceTest {

    @Test
    fun getValue_DelegatesToResolvers() {
        val root = Namespace.root()
        val thisProp = Name.of("this.prop")
        var resolvedNamespace: Namespace? = null
        var resolvedName: Name? = null
        val names = NameResolver { namespace, name ->
            resolvedNamespace = namespace
            resolvedName = name
            "THIS_PROP"
        }
        var resolvedKey: String? = null
        val values = ValueResolver { key ->
            resolvedKey = key
            "some_value"
        }

        val source = SimpleConfigurationSource(names, values)
        val value = source.getValue(root, thisProp)

        assertEquals("some_value", value)
        assertEquals(root, resolvedNamespace)
        assertEquals(thisProp, resolvedName)
        assertEquals("THIS_PROP", resolvedKey)
    }
}
